src/Runtime/Eval.hs

```diff
@@ -140,21 +129,19 @@ eval (B b) = return $ Vb b
 eval (S s) = return $ Vs s
 eval (Tuple es) = (sequence (map eval es)) >>= (return . Vt)
 eval (Ref n) = do
-  e <- ask
-  case lookup n e of
-        Just (Simple v e') -> newScope (e') (eval v)
+  e <- lookupName n
+  case e of
+        Just (v) -> return $ v
         _ -> return $ Err $ "Variable " ++ n ++ " undefined"
 eval (App n es) = do
-  args <- sequence $ map giveEnv es
+  args <- sequence $ map (eval) es
   f <- lookupName n
   case f of
-    Just (Fun params e env') -> newScope ((zip params args) ++ env') (eval e) -- FIXME
-    Nothing -> case lookup n builtins of
-        Just (f) -> f es
-        Nothing -> return $ Err $ "Couldn't find " ++ n ++ "in enviroment!"
+    Just (Vf params env' e) -> newScope ((zip params args) ++ env') (eval e) -- FIXME
```

Comment:
If FIXME means there's still a problem here, we can probably open up an issue to address it when this in.

Reply:
oops, left that in. I originally thought I looked messy, so I implemented that newScope helper function which abstracts away a little bit of the ReaderT plumbing